keps/sig-api-machinery/4460-per-request-deadline/README.md

```diff
@@ -0,0 +1,1073 @@
+			RequestAuditID: "", HandlerDoneCh: handlerDoneCh}
+		tracker.Enter(data)
+	})
+
+	defer stopFn()
+	handler.ServeHTTP(w, req)
+```
+
+
+The `PostStartHook` goroutine will wake up every `5m`, and invoke `Sweep` which
+will  perform the following steps on every `RequestData` item:
+
+- if the `HandlerDoneCh` channel is closed, it indicates that the request handler
+  returned after the context had timed out. The entry will be removed from the
+  list and post-timeout activity will be reported in the metrics/logs.
+- If an entry is resident for more than `1h`, we report it as hanging "forever"
+  and remove it from the list, so the list does not grow infinitely.
+
+> These two intervals can be tweaked as needed
+
+Additionally, we can expose a debugging endpoint in the kube-apiserver that 
+dumps a live snapshot of the post-timeout list.
+
+### Impact on the Client
+
+It's important to understand how the the client is impacted with this feature enabled; 
+upon timeout of a request, what the client sees depends on the following factors:
+
+- a) Protocol of the request, HTTP/1x or HTTP/2.0.
+- b) Whether the client specifies a timeout in the `timeout` parameter of the request.
+- c) Whether the request handler writes to the `ResponseWriter` object by invoking
+  `Write`, before the deadline exceeds.
+- d) Whether the request handler invokes `Flush` on the `ResponseWriter` object, 
+  preceded  by a `Write`, before the deadline exceeds.
+
+If enabling this feature introduces a change in what the client observes, then it should be 
+noted, and must be documented with a test. Let's outline the current behavior today, with
+the `timeout` filter enabled:
+
+When a request times out, the client today receives a `504 GatewayTimeout` only if the
+request handler does not write to the underlying `ResponseWriter` object before
+timeout occurs.
+
+Moreover, if the client sets a timeout value in the request URI `{host/path}?timeout=10s` 
+it will yield the same deadline for the request on both the client and the server. 
+This is the case with `client-go` today. In this scenario, if a request times out, 
+the client is most likely to receive a `context deadline exceeded` error due to
+the fact that the client will see the `Context` of the client request be expired
+well before the `504 GatewayTimeout` response from the server makes the round-trip
+to the client.
+
+Also, the client can not control when or whether the request handler writes to
+the `ResponseWriter` object. So, today a client can not deterministically catch
+the `504 GatewayTimeout` error from the server when a request times out.
+
+With the enablement of this feature, and the removal of the `timeout` filter the
+apiserver is no longer sending a `504 GatewayTimeout` response when a request 
+times out.
+
+The following goes through different scenarios, and for each we describe how the
+server and client behaves differently when this feature is enabled.
+
```

Comment:
What about scenarios where the request handler writes and flushes before the deadline, and does not write or flush after the deadline, but does not return until after the deadline. With this feature disabled, the client will get an error, and with this feature enabled, the client will not get an error. Have I got that right?

Reply:
for http2 (i am assuming the client does not enforce any timeout on its side)
- with this feature enabled: the client receives a `stream reset` error as soon as the write deadline exceeds, even though the handler is still executing, or is blocked.
- with this feature disabled: if the feature is disabled, then the `timeout` filter is in use, and the `timeout` filter, while it tries to write `504` to the client, it will notice that the `ResponseWriter` object has already been written to, and so it will panic. The client will see a `stream reset` error as a result of the panic from the server. 

https://github.com/kubernetes/kubernetes/blob/9791f0d1f39f3f1e0796add7833c1059325d5098/staging/src/k8s.io/apiserver/pkg/server/filters/timeout.go#L262-L276